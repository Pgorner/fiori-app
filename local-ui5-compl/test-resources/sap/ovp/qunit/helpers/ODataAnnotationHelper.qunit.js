sap.ui.define([
    "sap/ovp/helpers/ODataAnnotationHelper"
], function(ODataAnnotationHelper) {
    "use strict";

    QUnit.test("ODataAnnotationHelper - getRecords, get the annotation records for an entity type", function (assert) {
        var sAnnotationPath = "com.sap.vocabularies.UI.v1.Identification#navigationIntentBased";
        var aRecords = [
            {
                "SemanticObject":{"String":"browse"},
                "Action":{"String":"books1"},
                "RecordType":sAnnotationPath
            }
        ];
        var oEntityType = {
            "name": "Books",
            "com.sap.vocabularies.UI.v1.Identification#navigationIntentBased": aRecords
        };
        var oModel = {};
        assert.ok(ODataAnnotationHelper.getRecords(oEntityType, sAnnotationPath, oModel) === aRecords, "annotation records are returned");
    });
});
