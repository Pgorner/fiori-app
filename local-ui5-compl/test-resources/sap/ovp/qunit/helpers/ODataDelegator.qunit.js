sap.ui.define([
    "sap/ovp/helpers/ODataDelegator",
    "sap/ovp/cards/MetadataAnalyser",
    "sap/ovp/helpers/V4/MetadataAnalyzer"
], function (
    ODataDelegator,
    V2MetadataAnalyser,
    V4MetadataAnalyser
) {
    "use strict";

    QUnit.test("ODataDelegator - getParametersByEntitySet, should call / delegate responsiblity to the appropriate handler based on model version", function (assert) {
        var oEntitySet = {
            "name": "Books",
            "entityType": "CatalogService.Books",
            "Org.OData.Capabilities.V1.SearchRestrictions": {"Searchable": {"Bool": "false"}}
        };
        var v2spy = sinon.stub(V2MetadataAnalyser, "getParametersByEntitySet").callsFake(function (){return true;});
        var v4spy = sinon.stub(V4MetadataAnalyser, "getParametersByEntitySet").callsFake(function (){return true;});

        //v2 model
        var oModel = {
            getODataVersion: function () {
                return "2.0";
            }
        };
        ODataDelegator.getParametersByEntitySet(oModel, oEntitySet);
        assert.ok(v2spy.called, "V2 metadata analyzer gets called");

        //V4 model
        oModel.getODataVersion = function () {
            return "4.0";
        }
        ODataDelegator.getParametersByEntitySet(oModel, oEntitySet);
        assert.ok(v4spy.called, "V4 metadata analyzer gets called");

    });

    QUnit.test("ODataDelegator - getPropertyFromEntityType, should call / delegate responsiblity to the appropriate handler based on model version", function (assert) {
        var sPropertyPath = "category";
        var oEntityType = {
            "name": "Books",
            "property": [
                {
                    "name": "amount",
                    "type": "Edm.Int32"
                },
                {
                    "name": "category",
                    "type": "Edm.String"
                }
            ]
        }
        var v2spy = sinon.stub(V2MetadataAnalyser, "getPropertyFromEntityType").callsFake(function (){return true;});
        var v4spy = sinon.stub(V4MetadataAnalyser, "getPropertyFromEntityType").callsFake(function (){return true;});

        //v2 model
        var oModel = {
            getODataVersion: function () {
                return "2.0";
            }
        };
        ODataDelegator.getPropertyFromEntityType(sPropertyPath, oEntityType, oModel);
        assert.ok(v2spy.called, "V2 metadata analyzer gets called");
        
        //V4 model
        oModel.getODataVersion = function () {
            return "4.0";
        }
        ODataDelegator.getPropertyFromEntityType(sPropertyPath, oEntityType, oModel);
        assert.ok(v4spy.called, "V4 metadata analyzer gets called");

    });
});
