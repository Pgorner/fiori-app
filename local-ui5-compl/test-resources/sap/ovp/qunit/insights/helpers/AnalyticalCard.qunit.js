sap.ui.define([
    "sap/ovp/insights/helpers/AnalyticalCard",
    "test-resources/sap/ovp/qunit/cards/utils",
    "test-resources/sap/ovp/mockservers",
    "sap/ovp/insights/helpers/Batch",
    "test-resources/sap/ovp/qunit/cards/charts/utils",
    "sap/ovp/cards/charts/VizAnnotationManager",
    "sap/ovp/helpers/ODataAnnotationHelper",
], function (
    AnalyticalCard,
    Utils,
    mockservers,
    BatchHelper,
    ChartUtils,
    OvpVizAnnotationManager,
    ODataAnnotationHelper,
) {
    "use strict";
    QUnit.module("sap.ovp.insights.helpers.AnalyticalCard", {
        beforeEach: function () {
            mockservers.loadMockServer(Utils.odataBaseUrl_salesShare, Utils.odataRootUrl_salesShare);
        },
        afterEach: function () {
            mockservers.close();
        },
    });

    QUnit.test("AnalyticalCardHelper - createChartContent - validate created chart content.", function (assert) {
        var cardTestData = {
            card: {
                id: "chart_28",
                model: "salesShare",
                template: "sap.ovp.cards.charts.analytical",
                settings: {
                    entitySet: "SalesShare",
                    selectionAnnotationPath: "com.sap.vocabularies.UI.v1.SelectionVariant#Eval_by_CtryCurr",
                    chartAnnotationPath: "com.sap.vocabularies.UI.v1.Chart#Eval_by_Currency_Test_Forecast",
                    presentationAnnotationPath: "com.sap.vocabularies.UI.v1.PresentationVariant#Eval_by_CtryCurr",
                    dataPointAnnotationPath: "com.sap.vocabularies.UI.v1.DataPoint#Eval_by_CtryCurr",
                    identificationAnnotationPath: "com.sap.vocabularies.UI.v1.Identification#Eval_by_CtryCurr",
                }
            },
            dataSource: {
                baseUrl: ChartUtils.odataBaseUrl,
                rootUri: ChartUtils.odataRootUrl,
                annoUri: ChartUtils.testBaseUrl + "data/salesshare/annotations.xml",

            }
        };
        var oModel = Utils.createCardModel(cardTestData);
        var fnDone = assert.async();
        Utils.createCardView(cardTestData, oModel).then(function (oView) {
            var oCardController = oView.getController();
            var oCardViz = oView.byId("analyticalChart");

            oCardController.vizFrame = oCardViz;
            var oHandler = oCardViz && oCardViz.getParent();
            var oController = {
                getView: function () {
                    return null;
                },
                getOwnerComponent: function () {
                    return null;
                },
            };
            OvpVizAnnotationManager.buildVizAttributes(oCardViz, oHandler, oController);

            oView.getController().oMainComponent = {
                getGlobalFilter: function () {
                    return null;
                }
            };
            oView.getController().oCardComponentData.mainComponent = oView.getController().oMainComponent;
            var oCardComponentData = oCardController.oCardComponentData;
            var oCardDefinition = {
                vizFrame: oCardController.vizFrame,
                entitySet: "SalesShareDonut",
                entityType: oCardController.getEntityType(),
                cardComponentName: "Analytical",
                cardComponentData: oCardComponentData,
                cardComponent: oCardController.oCardComponent,
                view: oView
            };

            var oSapCard = {
                configuration: {
                    parameters: {}
                }
            };
            var getRequestUrlAndEntityInfoStub = sinon.stub(BatchHelper, "_getRequestUrlAndEntityInfo").returns({
                sPath: "SalesShareColumnStacked?$orderby=Region%20asc",
                _entitySet: {
                    value: "SalesShareColumnStacked"
                },
                _urlSuffix: {
                    value: ""
                }
            });
            var oEntityModel = oView.getModel();
            var oSettings = oCardDefinition.cardComponentData.settings;

            var oChartAnnotation = ODataAnnotationHelper.getRecords(oCardDefinition.entityType, oSettings.chartAnnotationPath, oEntityModel);
            var sTitle = oChartAnnotation.Title && oChartAnnotation.Title.String;

            var oChartContent = AnalyticalCard.createChartContent(oCardDefinition, oSettings.chartAnnotationPath, oSapCard);

            assert.deepEqual(oChartContent.chartType, oCardViz.getVizType(), "checking chart-type value");
            assert.deepEqual(oChartContent.title.text, sTitle, "checking for chart title value");
            assert.ok(
                oChartContent.title.visible == true,
                "Check if title is visible"
            );
            assert.deepEqual(oChartContent.dimensions.length, 1, "chart is having dimension as per configuration");

            assert.deepEqual(oChartContent.measures.length, oCardViz.getDataset().getMeasures().length, "chart is having measures as per configuration");
            assert.deepEqual(oChartContent.feeds.length, oCardViz.getFeeds().length, "chart is having feeds as per configuration");

            assert.deepEqual(oChartContent.dimensions[0].label, oCardViz.getDataset().getDimensions()[0].getName(), "checking dimensions value according to the entityset");
            assert.ok(
                oChartContent.measures[2].label === oCardViz.getDataset().getMeasures()[2].getName(),
                "Check if measure is being considered"
            );
            assert.ok(
                oCardViz.getVizProperties()["legend"].itemMargin === 1.25,
                "legends are set with the correct item margin"
            );
            getRequestUrlAndEntityInfoStub.restore();

            fnDone();
        });
    })
});
