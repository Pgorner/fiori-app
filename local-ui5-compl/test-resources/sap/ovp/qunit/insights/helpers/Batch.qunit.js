sap.ui.define([
    "sap/ovp/insights/helpers/Batch",
    "test-resources/sap/ovp/qunit/cards/utils",
    "test-resources/sap/ovp/mockservers",
    "sap/ovp/cards/AnnotationHelper" // This is needed for the qunits to run
], function(
    BatchHelper,
    utils,
    mockservers,
    CardAnnotationHelper
) {
    "use strict";

    QUnit.test("BatchHelper Utility - _removeQueryParam, remove requested query parameter from card binding url when there are multiple parameters", function (assert) {
        var sUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&_requestFrom=ovp_internal";
        var sExpectedUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY";
        assert.ok(BatchHelper._removeQueryParam(sUrl, "_requestFrom") === sExpectedUrl, "_requestFrom parameter is removed from url parameters");
    });

    QUnit.test("BatchHelper Utility - _removeQueryParam, remove requested query parameter from card binding url when there is single parameter", function (assert) {
        var sUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?_requestFrom=ovp_internal";
        var sExpectedUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked";
        assert.ok(BatchHelper._removeQueryParam(sUrl, "_requestFrom") === sExpectedUrl, "_requestFrom parameter is removed from url parameters");
    });

    QUnit.test("BatchHelper Utility - _removeQueryParam, remove requested query parameter from card binding url when there are multiple parameters, remove first parameter and correct ? and & position", function (assert) {
        var sUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&_requestFrom=ovp_internal";
        var sExpectedUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?_requestFrom=ovp_internal";
        assert.ok(BatchHelper._removeQueryParam(sUrl, "$select") === sExpectedUrl, "$select parameter is removed from url parameters");
    });

    QUnit.test("BatchHelper Utility - _getQueryParam, get requested query parameter(format=>key=value)", function (assert) {
        var sUrl = "/sap/smartbusinessdemo/services/SalesShare.xsodata/SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&_requestFrom=ovp_internal";
        assert.ok(BatchHelper._getQueryParam(sUrl, "$select") === "$select=Sales%2cSales_CURRENCY", "Received the correct value for $select");
    });

    QUnit.test("BatchHelper - enhanceHeaderAndContentURL, _headerDataUrl and _contentDataUrl are unchanged when _mandatoryODataParameters and _mandatoryODataFilters are both empty", function (assert) {
        var oCardConfig = {
            "configuration": {
                "parameters": {
                    "_mandatoryODataParameters": { 
                        "value": [] 
                    },
                    "_mandatoryODataFilters": { 
                        "value": [] 
                    },
                    "_headerDataUrl": {
                        "value": "SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&$skip=0&$top=1"
                    },
                    "_contentDataUrl": {
                        "value": "SalesShareColumnStacked?$orderby=Region%20asc&$select=StatusCriticality%2cRegion%2cTotalSales%2cTotalSales_CURRENCY&$skip=0&$top=20"
                    }
                }
            }
        };
        var sExpectedHeaderDataUrl = "SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&$skip=0&$top=1";
        var sExpectedContentDataUrl = "SalesShareColumnStacked?$orderby=Region%20asc&$select=StatusCriticality%2cRegion%2cTotalSales%2cTotalSales_CURRENCY&$skip=0&$top=20"
        BatchHelper.enhanceHeaderAndContentURL(oCardConfig);
        assert.ok(oCardConfig.configuration.parameters._headerDataUrl.value === sExpectedHeaderDataUrl, "Header Data Url should be unchanged");
        assert.ok(oCardConfig.configuration.parameters._contentDataUrl.value === sExpectedContentDataUrl, "Content Data Url should be unchanged");
    });

    QUnit.test("BatchHelper - enhanceHeaderAndContentURL, _headerDataUrl and _contentDataUrl should be set to empty string when one of _mandatoryODataParameters or _mandatoryODataFilters has a value but associated parameter values are empty, _mandatoryODataParameters value is empty", function (assert) {
        var oCardConfig = {
            "configuration": {
                "parameters": {
                    "P_ExchangeRateType": {
                        "value": "",
                        "type": "string"
                    },
                    "_mandatoryODataParameters": { 
                        "value": [
                            "P_ExchangeRateType"
                        ] 
                    },
                    "_mandatoryODataFilters": { 
                        "value": [] 
                    },
                    "_headerDataUrl": {
                        "value": "SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&$skip=0&$top=1"
                    },
                    "_contentDataUrl": {
                        "value": "SalesShareColumnStacked?$orderby=Region%20asc&$select=StatusCriticality%2cRegion%2cTotalSales%2cTotalSales_CURRENCY&$skip=0&$top=20"
                    }
                }
            }
        };
        BatchHelper.enhanceHeaderAndContentURL(oCardConfig);
        assert.ok(oCardConfig.configuration.parameters._headerDataUrl.value === "", "Header Data Url should be empty");
        assert.ok(oCardConfig.configuration.parameters._contentDataUrl.value === "", "Content Data Url should be empty");
    });

    QUnit.test("BatchHelper - enhanceHeaderAndContentURL, _headerDataUrl and _contentDataUrl should be set to empty string when one of _mandatoryODataParameters or _mandatoryODataFilters has a value but associated parameter values are empty, _mandatoryODataFilters value is empty", function (assert) {
        var oCardConfig = {
            "configuration": {
                "parameters": {
                    "ExchangeRateType": {
                        "value": "{\"Parameters\":[],\"SelectOptions\":[{\"PropertyName\":\"ExchangeRateType\",\"Ranges\":[]}]}",
                        "type": "string"
                    },
                    "_mandatoryODataParameters": { 
                        "value": [] 
                    },
                    "_mandatoryODataFilters": { 
                        "value": [
                            "ExchangeRateType"
                        ] 
                    },
                    "_headerDataUrl": {
                        "value": "SalesShareColumnStacked?$select=Sales%2cSales_CURRENCY&$skip=0&$top=1"
                    },
                    "_contentDataUrl": {
                        "value": "SalesShareColumnStacked?$orderby=Region%20asc&$select=StatusCriticality%2cRegion%2cTotalSales%2cTotalSales_CURRENCY&$skip=0&$top=20"
                    }
                }
            }
        };
        BatchHelper.enhanceHeaderAndContentURL(oCardConfig);
        assert.ok(oCardConfig.configuration.parameters._headerDataUrl.value === "", "Header Data Url should be empty");
        assert.ok(oCardConfig.configuration.parameters._contentDataUrl.value === "", "Content Data Url should be empty");
    });

    QUnit.test("BatchHelper - getBatchObject, Add _contentDataUrl to sap.card configuration, when KPI Annotation is missing", function (assert) {
        var oCardDefinition = {
            "entitySet": {
                "name": "SalesShareColumnStacked"
            },
            "cardComponentData": {
                "settings": {
                    title: "Sales by Country and Region",
                    subTitle: "Sales by Country and Region",

                }
            }
        };
        var oConfigParams = {
            parameters: {}
        };
        var getRequestUrlAndEntityInfoStub = sinon.stub(BatchHelper, "_getRequestUrlAndEntityInfo").returns({
            sPath: "SalesShareColumnStacked?$orderby=Region%20asc",
            _entitySet: {
                value: "SalesShareColumnStacked"
            },
            _urlSuffix: {
                value: ""
            }
        });

        var sExpectedContentDataUrl = "SalesShareColumnStacked?$orderby=Region%20asc";
        var oBatch= BatchHelper.getBatchObject(oCardDefinition, oConfigParams);
        assert.equal(oConfigParams.parameters._contentDataUrl.value, sExpectedContentDataUrl, "Content Data Url value is set");
        assert.equal(oBatch.content.url, sExpectedContentDataUrl, "Batch.content.url value is set")
        getRequestUrlAndEntityInfoStub.restore();
    });

    QUnit.test("BatchHelper - getBatchRequestPath - returns the correct path for a batch call", function (assert) {
        var sPath = "/content/d/results",
            sV4RequestPath = "/content/value",
            oEntityModel = {
                getODataVersion: function () {
                    return "2.0";
                }
            };
        assert.ok(BatchHelper.getBatchRequestPath(sPath, oEntityModel) === sPath, "No change in case of V2");

        oEntityModel.getODataVersion = function () {
            return "4.0";
        };
        assert.ok(BatchHelper.getBatchRequestPath(sPath, oEntityModel) === sV4RequestPath, "correct batch request path is returned");
    });

    QUnit.test("BatchHelper - getBatchResultCount - returns the correct count for a batch call", function (assert) {
        var sCountPath = "/content/d/__count",
            sV4RequestCountPath = "/content/@odata.count",
            oEntityModel = {
                getODataVersion: function () {
                    return "2.0";
                }
            };
        assert.ok(BatchHelper.getBatchResultCount(sCountPath, oEntityModel) === sCountPath, "No change in case of V2");

        oEntityModel.getODataVersion = function () {
            return "4.0";
        };
        assert.ok(BatchHelper.getBatchResultCount(sCountPath, oEntityModel) === sV4RequestCountPath, "correct batch request count is returned");
    });

    QUnit.test("BatchHelper - _getRequestUrlAndEntityInfo - validate path and entity info for Table card", function (assert) {
        mockservers.loadMockServer(utils.odataBaseUrl_salesOrder, utils.odataRootUrl_salesOrder);
        var cardTestData = {
            card: {
                id: "card_01",
                model: "salesOrder",
                template: "sap.ovp.cards.table",
                settings: {
                    title: "New Sales Orders",
                    subTitle: "Today",
                    entitySet: "SalesOrderSet",
                    category: "Sales Order Line Items - Table"
                }
            },
            dataSource: {
                baseUrl: utils.odataBaseUrl_salesOrder,
                rootUri: utils.odataRootUrl_salesOrder,
                annoUri: utils.testBaseUrl + "data/annotations.xml",
            }
        };

        var oModel = utils.createCardModel(cardTestData);
        var fnDone = assert.async();

        utils.createCardView(cardTestData, oModel).then(function (oView) {
            var oController = oView.getController();
            var oItemsBinding = oController.getCardItemsBinding();
            oController.onAfterRendering();
            //CreateData Change event
            oItemsBinding.fireDataReceived();
            var itemBindingInfo = oView.byId('ovpTable').getBindingInfo('items');
            oView.getController().oMainComponent = {
                getGlobalFilter: function() {
                    return null;
                }
            };
            oView.getController().oCardComponentData.mainComponent = oView.getController().oMainComponent;
            var oCardDefinition = {
                entitySet: oController.entitySet,
                entityType: oController.entityType,
                cardComponentName: "Table",
                cardComponentData: oController.oCardComponentData,
                cardComponent: oController.oCardComponent,
                itemBindingInfo: itemBindingInfo,
                view: oView
            };
            var oConfigParams = {
                parameters : {
                    _mandatoryODataFilters: {
                        value: []
                    }
                }
            };
            var oResult = {
                "sPath": "SalesOrderSet?$inlinecount=allpages&$skip=0&$top=13",
                "_entitySet": {
                    "value": "SalesOrderSet"
                },
                "_urlSuffix": {
                    "value": ""
                }
            };
            assert.deepEqual(BatchHelper._getRequestUrlAndEntityInfo(oCardDefinition, oConfigParams, false), oResult, "Returns the correct URL and entity for the card");

            fnDone();
            mockservers.close();
        });
    });

});
