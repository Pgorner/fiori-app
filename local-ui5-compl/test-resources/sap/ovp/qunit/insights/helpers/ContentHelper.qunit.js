sap.ui.define([
    "sap/ovp/insights/helpers/ContentHelper"
], function(ContentHelper) {
    "use strict";

    QUnit.test("ContentHelper  - function: addBrackets", function (assert) {
        assert.ok(ContentHelper.addBrackets("pathName") === "{pathName}", "Bracket is added to a path.");
        assert.ok(ContentHelper.addBrackets("{pathName}") === "{pathName}", "Additional unwanted bracket is not added to a path.");
        assert.ok(ContentHelper.addBrackets("{= ${pathName}}") === "{= ${pathName}}", "Additional unwanted bracket is not added to an expression.");
    });

    QUnit.test("ContentHelper  - function: getHeaderStatus", function (assert) {
        var oBatch;
        assert.deepEqual(ContentHelper.getHeaderStatus(oBatch), { text: "{= ${/d/__count} === '0' ? '' : extension.formatters.formatHeaderCount( ${/d/__count}) }" }, "Header status with count when there is call without batch.");
        oBatch = {batch: true};
        assert.deepEqual(ContentHelper.getHeaderStatus(oBatch), { "text": "{= ${/content/d/__count} === '0' ? '' : extension.formatters.formatHeaderCount( ${/content/d/__count}) }" }, "Header status with count when there is call with batch.");
    });

    QUnit.test("ContentHelper  - function: getData", function (assert) {
        var oBatch;
        assert.deepEqual(ContentHelper.getData(oBatch), { path: "/d/results" }, "Data path when there is call without batch.");
        oBatch = {batch: true};
        assert.deepEqual(ContentHelper.getData(oBatch), { path: "/content/d/results" }, "Data path when there is call with batch.");
    });

    QUnit.test("ContentHelper  - function: isAnExpression", function (assert) {
        assert.equal(ContentHelper.isAnExpression("{= ${pathName}}"), true, "String is an expression.");
        assert.equal(ContentHelper.isAnExpression("{pathName}"), false, "Path String is not an expression.");
        assert.equal(ContentHelper.isAnExpression("pathName"), false, "String is not an expression.");
    });

    QUnit.test("ContentHelper  - function: buildValueStateExpression", function (assert) {
        var sStateType = "color";
        assert.equal(ContentHelper.buildValueStateExpression("{= ${pathName}}"), "{= ${pathName}}", "Returns expression directly when expression is already present.");
        assert.equal(ContentHelper.buildValueStateExpression("Good", sStateType), "{= extension.formatters.formatCriticality('Good', 'color')}", "Returns new expression with value.");
        assert.equal(ContentHelper.buildValueStateExpression("{pathName}", sStateType), "{= extension.formatters.formatCriticality(${pathName}, 'color')}", "Returns new expression with path variable.");
    });

    QUnit.test("ContentHelper  - function: getBindingPathOrValue", function (assert) {
        // check empty
        var sPropertyName, sExpectedExpression;
        var oProperties = {
                property1: "property1Value", // vaue
                property2: "{property2Path}", // single path binding
                property3: {
                    parts: [{ path : '{property3Path}'}]
                },
                property4: {
                    parts: [{ path : '{property4Path}'}],
                    type: {
                        oFormat: {
                            oFormatOptions: {
                                "percentSign": "%",
                                "pattern": "#,##0.###"
                            }
                        }
                    }
                },
                property5: {
                    parts: [{ path : '{property5Path}'},
                        {
                            "value": {
                                "dateFormat": {
                                    "relative": true,
                                    "relativeScale": "auto",
                                    "bUTC": false
                                },
                                "bUTC": false,
                                "functionName": "CardAnnotationhelper.formatDate"
                            },
                            "model": "ovpCardProperties",
                            "mode": "OneWay"
                        }],
                    formatter: function() {}
                },
                property6: {
                    parts: [{ path : '{property6Path1}'}, { path : '{property6Path2}'},
                        {
                            "value": {
                                "numberOfFractionalDigits": "0",
                                "scaleFactor": "1000",
                                "functionName": "CardAnnotationhelper.formatCurrency"
                            },
                            "model": "ovpCardProperties",
                            "mode": "OneWay"
                        }],
                    formatter: function() {}
                },
                property7: {
                    parts: [{ path : '{property7Path1}'}, { path : '{property7Path2}'},
                        {
                            "value": {
                                "numberOfFractionalDigits": 1,
                                "functionName": "CardAnnotationhelper.formatNumberCalculation"
                            },
                            "model": "ovpCardProperties",
                            "mode": "OneWay"
                        }],
                    formatter: function() {}
                }
            };
        var oControl = {
                getProperty: function (propertyName) {
                    return oProperties[propertyName];
                },
                getBindingInfo: function () {
                    return;
                },
                getBindingPath: function (propertyName) {
                    return oProperties[propertyName];
                }
            };
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), null, "Returned null when parameters are empty.");
        // check direct value
        sPropertyName = "property1";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), "property1Value", "Returned property value when binding is not present.");
        // check simple binding
        oControl.getBindingInfo = function (propertyName) {
            return oProperties[propertyName];
        };
        sPropertyName = "property2";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), "{property2Path}", "Returned property path when simple binding is present.");
        // check binding with parts
        sPropertyName = "property3";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), "{property3Path}", "Returned property path when single path is present but in parts.");

        // check binding with parts and type with formatter
        sPropertyName = "property4";
        sExpectedExpression = "{= extension.formatters.formatNumber({'percentSign':'%','pattern':'#,##0.###'},[0],${{property4Path}})}";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), sExpectedExpression, "Returned expression with formatting from 'type'.");

        // date formatter
        sPropertyName = "property5";
        sExpectedExpression = "{= extension.formatters.formatDate(${{property5Path}}, {'relative':true,'relativeScale':'auto','bUTC':false})}";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), sExpectedExpression, "Returned expression with date formatter");

        // currency formatter
        sPropertyName = "property6";
        sExpectedExpression = "{= extension.formatters.formatCurrency({'numberOfFractionalDigits':'0','scaleFactor':'1000'},false,${{property6Path1}},${{property6Path2}})}";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), sExpectedExpression, "Returned expression with currency formatter");

        // formatNumberCalculation formatter
        sPropertyName = "property7";
        sExpectedExpression = "{= extension.formatters.formatNumber({'numberOfFractionalDigits':1},[0],${{property7Path1}},${{property7Path2}})}";
        assert.equal(ContentHelper.getBindingPathOrValue(oControl, sPropertyName), sExpectedExpression, "Returned expression with formatNumberCalculation formatter");

    });  
});
