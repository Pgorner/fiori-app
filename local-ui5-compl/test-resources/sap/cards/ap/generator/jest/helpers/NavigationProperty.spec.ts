/*!
 * SAP UI development toolkit for HTML5 (SAPUI5)
 *  * (c) Copyright 2009-2025 SAP SE. All rights reserved
 */
import { ApplicationInfo } from "sap/cards/ap/generator/helpers/ApplicationInfo";
import { getNavigationPropertiesWithLabel } from "sap/cards/ap/generator/helpers/NavigationProperty";
import UIComponent from "sap/ui/core/UIComponent";

declare namespace sap {
	export namespace jest {
		function resolvePath(path: string): string;
	}
}

jest.mock(sap.jest.resolvePath("sap/cards/ap/generator/odata/ODataUtils"), () => {
	return {
		...jest.requireActual(sap.jest.resolvePath("sap/cards/ap/generator/odata/ODataUtils")),
		fetchDataAsync: jest.fn().mockResolvedValue({
			someNavigationProperty: {
				Status: "O",
				Status_Text: "Open"
			}
		}),
		getNavigationPropertyInfoFromEntity: jest.fn().mockReturnValue([
			{
				name: "someNavigationProperty",
				properties: [
					{
						label: "Lower Value",
						type: "Edm.String",
						name: "Status"
					},
					{
						label: "Confirmation",
						type: "Edm.String",
						name: "Status_Text"
					}
				]
			}
		])
	};
});

const sId = "testComponent";
const oManifest = {
	"sap.app": {
		id: sId,
		type: "application"
	}
};
const Component = UIComponent.extend("component", {
	metadata: {
		manifest: oManifest
	},
	createContent() {
		return null;
	}
});

const rootComponent = new Component(sId);
jest.spyOn(rootComponent, "getModel").mockImplementation(() => {
	return {
		sServiceUrl: "/sap/opu/odata",
		isA: () => false
	};
});

describe("getNavigationPropertiesWithLabel", () => {
	let windowSpy: jest.SpyInstance;

	beforeAll(() => {
		windowSpy = jest.spyOn(window, "window", "get");
		ApplicationInfo.createInstance(rootComponent);
	});

	afterAll(() => {
		windowSpy.mockRestore();
		ApplicationInfo.getInstance()._resetInstance();
	});

	test("should return properties with label and navigation property data", async () => {
		windowSpy.mockImplementation(() => ({
			hasher: {
				getHash: () => "test-intent&/testEntity(12345)"
			}
		}));

		const navigationProperty = "someNavigationProperty";
		const path = "/sap/opu/odata4/sap/c_salesordermanage_srv/srvd/sap/c_salesordermanage_sd/0001/";
		const result = await getNavigationPropertiesWithLabel(rootComponent, navigationProperty, path);
		expect(result).toMatchSnapshot();
	});
});
