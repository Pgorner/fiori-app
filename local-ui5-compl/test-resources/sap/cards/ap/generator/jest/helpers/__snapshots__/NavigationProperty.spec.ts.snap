// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNavigationPropertiesWithLabel should return properties with label and navigation property data 1`] = `
{
  "navigationPropertyData": {
    "someNavigationProperty": {
      "Status": "O",
      "Status_Text": "Open",
    },
  },
  "propertiesWithLabel": [
    {
      "label": "Lower Value",
      "labelWithValue": "Lower Value (O)",
      "name": "Status",
      "type": "Edm.String",
    },
    {
      "label": "Confirmation",
      "labelWithValue": "Confirmation (Open)",
      "name": "Status_Text",
      "type": "Edm.String",
    },
  ],
}
`;
